bolt/lib/Profile/StaleProfileMatching.cpp

```diff
@@ -592,10 +599,15 @@ void preprocessUnreachableBlocks(FlowFunction &Func) {
 /// Decide if stale profile matching can be applied for a given function.
 /// Currently we skip inference for (very) large instances and for instances
 /// having "unexpected" control flow (e.g., having no sink basic blocks).
-bool canApplyInference(const FlowFunction &Func) {
+bool canApplyInference(const FlowFunction &Func,
+                       const yaml::bolt::BinaryFunctionProfile &YamlBF) {
   if (Func.Blocks.size() > opts::StaleMatchingMaxFuncSize)
     return false;
 
+  if ((double)Func.MatchedExecCount / YamlBF.ExecCount >=
+      opts::MatchedProfileThreshold / 100.0)
+    return false;
```

Comment:
Trying to understand the rationale behind using dynamic counts to determine whether profile inference is safe. 

The way I see it is, we have two graph that we try to match, if we have many nodes in the graph that we have exact match, chances are higher that we can infer the correct match for the rest of the nodes. With that, we care about more how many nodes we can match statically. 

Say if we have 5 blocks with count distribution of 1M, 1K, 1K, 1k, 1K, if we have exact match for the 4 1K node (80% exact match), we should feel reasonably confident about inferring the remaining 1 node, even though if we look at counts, we have exact match for only <1%. 

WDYT?

Reply:
It's a tricky question how to define the cutoff in terms of sufficient matching.
I first thought of defining a block count based cutoff (if we matched >5% of blocks, proceed with matching), but then what if these are cold blocks covering <1% of exec count? In this case we'd end up guessing/propagating most samples.

For block-based matching, the threshold should be higher than 5%, perhaps closer to a half? For exec count based matching, I'd feel comfortable with 5% as threshold.

Reply:
> For block-based matching, the threshold should be higher than 5%, perhaps closer to a half?

Yes. Threshold of course need to be tuned based on the heuristic chosen. I just feel that block count based threshold could be a better proxy of how confident we are about the graph match and whether stale profile matching should proceed.. But I don't have very strong opinion.

Reply:
I’m leaning towards the block count heuristic now. I think the 1M and 4x1K exec count block case is likely pretty common – I imagine functions with loops would look a lot like this. Having some blocks matched exactly would suggest to me that there would likely be a reasonable amount of similarity between the profiled function and existing function relationally, which block coldness likely doesn’t have an outsized bearing on. I think having a reasonably high threshold for matched blocks would conservatively allow us to drop functions in high discrepancy – I’ll test this on a production binary.